### ISSUE TITLE
Alexandria package

### ISSUE THREAD

**Issue by mdedwards** [mdedwards/slippery-chicken#90]
Orm's CM 2.12 requires the Alexandria package and if you want to initially use asdf only, i.e. skip quicklisp, it's not at all clear which system that refers to and where it is to be found. I believe an asdf-only install of SC should be possible. Any ideas?

**Comment by rubenphilipp** [comment]
I think that we have several options here. One could be to encourage the (future) user to check the dependencies of sc's dependencies. In cm's .asd, alexandria is explicitly declared to be required (cf. https://github.com/ormf/cm/blob/a781d6c5a52ca04717648808c8fa139bba139477/cm.asd#L48). Or we could mention alexandria to be required in our readme (though, I think it would be more meaningful if we asked Orm to include such a hint in cm's readme). Alternatively, we could also add alexandria to sc's :depends-on list, though I think this could be more confusing as we don't use alexandria in sc. 
We could, of course, also add a form to the .asd which tests whether alexandria is available and throws a meaningful error, though this is – as far as I know – not common practice in ASDF.

**Comment by mdedwards** [comment]
I am not sure I’d go any of those routes to be honest. Up until now the app and the install script approach has made things very easy for those users for whom those worked straight out of the box. Ahem ;)

We need the ASD solution to be very simple too. so we simply need a few links to required packages and a few lines of code that the new user adds to the lisp initialisation file and basta. If that means relying on well established and secure packages from quick lisp  then so be it. By the way is common Music even available on quick lisp?

**Comment by mdedwards** [comment]
Honestly one of the attractions of asdf--once I'd got over my enduring reticence--is that it would replace such an install script. My preference would be to go with a wiki page that tells people where to get the packages they need (including the optional ones) and which lines to put in their lisp init. That way we remain more OS-agnostic.

**Comment by mdedwards** [comment]
That's fine with me!

**Comment by mdedwards** [comment]
Hi Ruben, as you've done most of the heavy lifting with asd, and as you're thus better informed as to what is necessary right now, would you be willing to go the extra mile and write install.md?

Right now my init file looks like this:

```
(require :asdf)
(load "~/ic/quicklisp/setup.lisp")
(ql:quickload 'cm)
(load "~/lisp/cmn/cmn-all.lisp")
(load "~/ic/code/lisp/clm-init.lsp")
;;; I use a symbolic link to put my repo (wherever it is) into ~/common-lisp
(asdf:load-system :slippery-chicken)
```

**Comment by rubenphilipp** [comment]
Sure, I will do this. Created an issue so I won't forget about this…

**Comment by rubenphilipp** [comment]
> Hi Ruben, as you've done most of the heavy lifting with asd, and as you're thus better informed as to what is necessary right now, would you be willing to go the extra mile and write install.md?
> 
> Right now my init file looks like this:
> 
> ```
> (require :asdf)
> (load "~/ic/quicklisp/setup.lisp")
> (ql:quickload 'cm)
> (load "~/lisp/cmn/cmn-all.lisp")
> (load "~/ic/code/lisp/clm-init.lsp")
> ;;; I use a symbolic link to put my repo (wherever it is) into ~/common-lisp
> (asdf:load-system :slippery-chicken)
> ```

By the way, you should also be able to load cmn (and clm, thus this seems to be more special in your case as I don't know what magic happens in clm-init.lsp) the way you load cm, since the package definitions of both also work for me. 
Here's how the relevant portion of my initfile looks like:

> (asdf:load-system :cmn)
> (asdf:load-system :cm)
> (asdf:load-system :clm)
> (asdf:load-system :slippery-chicken)
> (load "~/.sc-config.lisp")                   ; custom config

**Comment by mdedwards** [comment]
Yes, thanks Ruben, I remember you mentioning that and how it works for you but didn't for Leon. I guess this is now the time for me to try it, so I did and hit well-known dlopen errors:

```
  Error opening shared object "/Users/michael/Library/Mobile Documents/com~apple~CloudDocs/code/lisp/asdf/clm-5/libclm.so":
  dlopen(/Users/michael/Library/Mobile Documents/com~apple~CloudDocs/code/lisp/a
```

That's a recurring and frustrating problem with stuff I keep on iCloud: I can't believe the Apple geniuses decided on "Mobile Documents" as a directory name (i.e. with a space) knowing full well that many people are unix savvy and use systems that rely on the shell and as such keep hitting shell command problems because of that damned space (not to speak of the tildes).

Do you find that you have to compile clm and can *by hand* the first time or can you ask asdf to load a virgin system?

Also, I could not load :cm via asdf because it didn't find Alexandria: only ql worked for me there.

Best, Michael

**Comment by rubenphilipp** [comment]
Ah, I also more than one time stumbled upon this horrible directory. 

> Do you find that you have to compile clm and can by hand the first time or can you ask asdf to load a virgin system?

Yes, I did not have any problems to use asdf:load-system to load asdf – even on a fresh system – as the .asd does more or less the same as your `(load...` command (i.e. loading all.lisp). Of course, this is not really an ideal package definition but it works…

> Also, I could not load :cm via asdf because it didn't find Alexandria: only ql worked for me there.

I remember. I wonder why it works on my machine. Probably because alexandria is in my quicklisp cache which is – as far as I know – also considered in `asdf:load-system`.

**Comment by mdedwards** [comment]
> I remember. I wonder why it works on my machine. Probably because alexandria is in my quicklisp cache which is – as far as I know – also considered in asdf:load-system.

Ah OK. This is one of the reasons I was never keen on the whole QL/ASDF approach. It's a bit opaque. I don't know what it's loading from where.

**Comment by rubenphilipp** [comment]
Yes, I understand this. I will try to investigate this further when working on the installation instructions.
